### PR TITLE
fix 457 Support for Bearer Auth

### PR DESCRIPTION
**Pull request by davidwcarlson**
Fix proposed for Issue OpenAPITools#457
Similar to Issue OpenAPITools#1446 for typescript, Issue OpenAPITools#1577 for python

Specs defined as follows currently generate `HttpBasicAuth` and send an 
```Authorization: Basic [base64Encode(username + ":" + password)]``` header
```
    components:
      securitySchemes:
        bearer:
          type: http
          scheme: bearer
```

This change will generate code which uses a new `HttpBearerAuth` class, which will send an ```Authorization: [scheme] [accessToken]``` header.

This change is slightly larger and more impactful than reusing the `OAuth` when `CodegenSecurity.bearerBearer` is true, but it allows for scheme values other than bearer.

This fix was enabled by the recent commit of OpenAPITools@80ca67c

This PR is an alternative to PR OpenAPITools#1972

### PR checklist

- [X] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [X] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [X] Filed the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`~~, `3.4.x`, `4.0.x`~~. Default: `master`.
- [X] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR

(details of the change, additional tests that have been done, reference to the issue for tracking, etc)
Add a new Auth type for Bearer so that api specs using Bearer Authentication https://swagger.io/docs/specification/authentication/bearer-authentication/ use the correct auth (was BasicAuth).

Copying the Java technical committee as this is a change targeted to Java.
@bbdouglas  @JFCote  @sreeshas @jfiala  @lukoyanov @cbornet  @jeff9finger

### CODE REVIEW

**Comment by lorenzleutgeb** [comment]
What I am missing here is that both `Bearer` and `Basic` authentication are actually closely related and just two variants of HTTP authorization schemes. This connection is not really clear here. However it is a step in the right direction and probably a solution that generalises nicely can be put in place.

File: modules/openapi-generator/src/main/resources/Java/auth/BearerAuth.mustache [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  public String getBearerToken() {
+    return bearerToken;
+  }
+
+  public void setBearerToken(String bearerToken) {
+    this.bearerToken = bearerToken;
+  }
+
+  @Override
+  public void applyToParams(List<Pair> queryParams, Map<String, String> headerParams) {
```
Comment (wing328) [inline modules/openapi-generator/src/main/resources/Java/auth/BearerAuth.mustache]: Minor suggestions: adding docstrings to these public methods.

**Comment by mstead** [comment]
Any idea when this change will be released?

**Comment by davidwcarlson** [comment]
I am not a maintainer, but I expect it in the next release on Feb 28
according to this:
https://github.com/OpenAPITools/openapi-generator/issues/1462

**Comment by wing328** [comment]
I just posted an update in https://github.com/OpenAPITools/openapi-generator/issues/1462

**Comment by intelliot** [comment]
Is this supposed to generate Javadocs? I'm seeing this:
```2 warnings
[WARNING] Javadoc Warnings
[WARNING] /Users/user/company/project/client/java/src/main/java/com/company/auth/HttpBearerAuth.java:33: warning: no @return
[WARNING] public String getBearerToken() {
[WARNING] ^
[WARNING] /Users/user/company/project/client/java/src/main/java/com/company/auth/HttpBearerAuth.java:40: warning: no @param for bearerToken
[WARNING] public void setBearerToken(String bearerToken) {
[WARNING] ^
```

**Comment by intelliot** [comment]
Hmm. This doesn't seem to actually work: generating a Java SDK using master creates docs that say `Type: HTTP basic authentication`, and the code itself still returns an `HttpBasicAuth`:
```
HttpBasicAuth bearerAuth = (HttpBasicAuth) defaultClient.getAuthentication("bearerAuth");
```

**Comment by davidwcarlson** [comment]
What library are you specifying in your pom and generating code for?  

From what I can tell, there is going to be some separate work required for each library (to modify the mustache template for ApiClient for your library similar to [this code](https://github.com/OpenAPITools/openapi-generator/blob/db1b63780d57ffb5dd9a6d76123400c4e0788a52/modules/openapi-generator/src/main/resources/Java/libraries/jersey2/ApiClient.mustache#L90)).  

I admit was focused on trying to fix this for jersey2 for my own selfish purposes.  Maybe this will point you in the right direction to submit a pull request or maybe I can take a crack if I get some time.

I hope this helps!

**Comment by matthewbluezyoncom** [comment]
@intelliot 
I'm also creating a client for jersey2, but, I don't have an OAuth security scheme defined, only ApiKey and Http.
I will try and add an OAuth scheme to see if the methods are generated. This will enable me to test, but, alas I will not be able to proceed further until PR #2479 has been merged.
Many thanks for the prompt reply and the original work to add HttpBearerAuth 👏

**Comment by matthewbluezyoncom** [comment]
I added an OAuth scheme expecting to see the setBearerToken method HttpBearerAuth class.
Neither were generated.
I am going to stick with ApiKey and set the token via the key and Bearer scheme via the prefix

**Comment by wing328** [comment]
#2485 has been merged into master to provide better Bearer authentication support on all Java API clients. Please pull the latest master to give it a try.

**Pull request merged**